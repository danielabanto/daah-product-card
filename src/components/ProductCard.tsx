import styles from '../styles/styles.module.css'
import { useProducts } from '../hooks/useProducts'
import React,{ createContext, useContext } from 'react'
import { InitialValues, onChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/interfaces'
export interface ProductCardProps {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: ( args: ProductCardHandlers ) => JSX.Element;
  className?: string;
  style?: React.CSSProperties;
  value?: number;
  onChange?: ( args: onChangeArgs ) => void
  initialValues?: InitialValues;
}

const initalState: ProductContextProps = {
  counter: 0,
  increaseBy: () => null,
  product: { id: '', title: ''}
}

const ProductContext = createContext<ProductContextProps>(initalState)
const { Provider } = ProductContext

export const useProductContext = () => useContext(ProductContext)

export const ProductCard = ({ product, children, className="", style, value, onChange, initialValues }: ProductCardProps) => {
  const { 
    counter, 
    increaseBy, 
    maxCount,
    isMaxCountReached,
    reset
  } = useProducts({ onChange, product, value, initialValues })

  return (
    <Provider value={{
      counter,
      increaseBy,
      product,
      maxCount
    }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        { children({
          count: counter,
          isMaxCountReached,
          maxCount,
          product,
          increaseBy,
          reset,
        }) }
      </div>
    </Provider>
  )
}
