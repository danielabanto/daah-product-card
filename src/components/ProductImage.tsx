import { useProductContext } from "./ProductCard"
import noImage from '../assets/no-image.jpg'
import styles from '../styles/styles.module.css'
import React from 'react'

export interface Props { 
  className?: string;
  img?: string ; 
  style?: React.CSSProperties; 
}

export const ProductImage = ({ img, className='', style }: Props ) => {
  const { product: { img : image } } = useProductContext()
  return (
    <img 
      className={ `${styles.productImg} ${className}` } 
      src={ img || image || noImage } 
      alt='Product' 
      style={style} 
    />
  )
}