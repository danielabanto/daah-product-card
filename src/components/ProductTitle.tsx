import { useProductContext } from "./ProductCard"
import styles from '../styles/styles.module.css'
import React from 'react'

export interface Props { 
  className?: string;
  title?: string;
  style?: React.CSSProperties;
}

export const ProductTitle = ({ title, className="", style }: Props) => {
  const { product: { title : productTitle} } = useProductContext()
  return (
      <span className={ `${styles.productDescription} ${className}` } style={style}>
        { title || productTitle }
      </span>
  )
}