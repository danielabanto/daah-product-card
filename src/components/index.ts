import { ProductCard as PC } from './ProductCard'
import { ProductButtons } from './ProductButtons'
import { ProductImage } from './ProductImage'
import { ProductTitle } from './ProductTitle'
import { ProductCardHOCProps } from '../interfaces/interfaces'

export const ProductCard: ProductCardHOCProps = Object.assign( PC,{
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
})


// En caso de querer la otra forma

export {ProductTitle}
export {ProductImage}
export {ProductButtons}