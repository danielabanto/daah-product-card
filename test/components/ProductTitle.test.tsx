import React from 'react';
import renderer from 'react-test-renderer'
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('debe mostrar titutlo personalizado', () => {
    const wrapper = renderer.create(
      <ProductTitle title='Custom Title' />
    )
    expect(wrapper.toJSON()).toMatchSnapshot()
  });

  test('debe mostrar el nombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => (
          <ProductTitle />
        )}
      </ProductCard>
    )
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
});
