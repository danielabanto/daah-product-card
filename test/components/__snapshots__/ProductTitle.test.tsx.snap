// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe mostrar el nombre del producto 1`] = `
<div
  className="productCard "
>
  <span
    className="productDescription "
  >
    Coffee Mug - Sin imagen
  </span>
</div>
`;

exports[`ProductTitle debe mostrar titutlo personalizado 1`] = `
<span
  className="productDescription "
>
  Custom Title
</span>
`;
